ds4drv/__main__.py

```diff
@@ -108,26 +109,55 @@ def read_report(self):
     def run(self):
         self.loop.run()
 
-    def exit(self, *args):
+    def exit(self, *args, error = True):
         if self.device:
             self.cleanup_device()
 
-        self.logger.error(*args)
-        self.error = True
+        if error == True:
+            self.logger.error(*args)
+            self.error = True
+        else:
+            self.logger.info(*args)
 
 
 def create_controller_thread(index, controller_options, dynamic=False):
     controller = DS4Controller(index, controller_options, dynamic=dynamic)
 
     thread = Thread(target=controller.run)
-    thread.daemon = True
     thread.controller = controller
     thread.start()
 
     return thread
 
 
+class SigintHandler(object):
+
+    def __init__(self, threads):
+        self.threads = threads
+
+    def cleanup_controller_threads(self):
+        for thread in self.threads:
+            thread.controller.exit("Cleaning up...", error=False)
+            thread.controller.loop.stop()
+            thread.join()
+
+    def __call__(self, signum, frame):
+        old_handler = signal.getsignal(signum)
+        signal.signal(signum, signal.SIG_DFL)
+
+        self.cleanup_controller_threads()
+
+        signal.signal(signum, old_handler)
```

Comment:
`signal.signal(signum, self)` isn't necessarily the same. I've considered the case where someone has taken our sigint handler and wrapped it in something else, or maybe the sigint handler method has been called manually. Probably overkill.

I restore the signal handler because I have no idea what exactly happens during `sys.exit(0)`. It might be better to set it to the default in case something hangs and the user wants to spam ^C to get it to close.